src/vision/status_quo/alan_tries_a_socket_sink.md

```diff
@@ -0,0 +1,106 @@
+
+However, as he's working, Alan encounters a situation where the socket needs to be written to within an async thread, and the traits just aren't working. He wants to split the stream into a sender and receiver:
+
+```rust
+use futures::{SinkExt, StreamExt};
+use async_std::sync::{Arc, Mutex};
+use log::{debug, info, warn};
+
+/// In the connection handler:
+let (ws_sender, mut ws_receiver) = ws_stream.split();
+let ws_sender = Arc::new(Mutex::new(ws_sender));
+
+while let Some(msg) = ws_receiver.next().await {
+    debug!("Received new WS RPC message: {:?}", msg);
+
+    // Echo the request:
+    match ws_sender.lock().await.send_string(msg).await {
+        Ok(_) => info!("New WS data sent."),
+        Err(_) => warn!("WS connection closed."),
+    };
+}
+```
+
+This is necessary because both sides of the stream are stateful iterators, and if they're kept as one, the lock on the receiver can't be released within the loop. He's seen this pattern used in other projects in the Rust community, but is frustrated to find that the `Sink` trait wasn't implemented in the WebSockets middleware library he's using.
+
+Alan also tries creating a sort of poller worker thread using an intermediary messaging channel, but he has trouble reasoning about the code and wasn't able to get it to compile:
+
+```rust
+use async_std::channel;
+use async_std::sync::{Arc, Mutex};
+use log::{debug, info, warn};
+
+/// In the connection handler:
+let (ws_sender, mut ws_receiver) = channel::unbounded::<String>();
+let ws_receiver = Arc::new(ws_receiver);
+
+let ws_stream = Arc::new(Mutex::new(ws_stream));
+let poller_ws_stream = ws_stream.clone();
+
+async_std::task::spawn(async move {
+    while let Some(msg) = ws_receiver.next().await {
+        match poller_ws_stream.lock().await.send_string(msg).await {
+            Ok(msg) => info!("New WS data sent. {:?}", msg),
+            Err(msg) => warn!("WS connection closed. {:?}", msg),
+        };
+    }
+});
+
+while let Some(msg) = ws_stream.lock().await.next().await {
+    // Echo the request?
+    ws_sender.send(msg);
+}
+```
+
+Alan wonders if he's thinking about it wrong, but the solution isn't as obvious as his earlier `Sink` approach. Looking around, he realizes a solution to his problems already exists-- as others have been in his shoes before-- within two other nearly-identical pull requests, but they were both closed by the project maintainers. He tries opening a third one with the same code, pointing to an example where it was actually found to be useful. To his joy, his original approach works with the code in the closed pull requests in his local copy! Alan's branch is able to compile for the first time.
+
+However, almost immediately, his request was shut down like the others, and the maintainer suggests he try the complex, unobvious, and unspecific solution that he had already tried and just couldn't get it to work.
```

Comment:
If it's not too long, is it possible to have a snippet of code that shows the code that Alan finds complex/unobvious/unspecific?

Reply:
That would indeed be very helpful

Reply:
That was what the second example was for, I'm not sure if there's a better way to explain that?